### PR TITLE
Use NodeId/HirId instead of DefId for local variables.

### PR DESCRIPTION
**Pull request by eddyb**
r? @michaelwoerister

### CODE REVIEW

File: src/librustc_save_analysis/lib.rs [hunk 1]
```diff
@@ -1013,7 +1020,15 @@ fn id_from_def_id(id: DefId) -> rls_data::Id {
 
 fn id_from_node_id(id: NodeId, scx: &SaveContext) -> rls_data::Id {
     let def_id = scx.tcx.hir.opt_local_def_id(id);
-    def_id.map(|id| id_from_def_id(id)).unwrap_or_else(null_id)
+    def_id.map(|id| id_from_def_id(id)).unwrap_or_else(|| {
+        // Create a *fake* `DefId` out of a `NodeId` by subtracting the `NodeId`
+        // out of the maximum u32 value. This will work unless you have *billions*
+        // of definitions in a single crate (very unlikely to actually happen).
```
Comment (michaelwoerister) [inline src/librustc_save_analysis/lib.rs]: OK.. `:D` 

`rls_data::Id` should arguably be switched to a `(DefPathHash, LocalId)` anyway at some point.

**Review by michaelwoerister — COMMENTED** [review]
Nice! It's a bit unfortunate that `Def::def_id()` can't be used as universally anymore but that's OK. Having fewer DefIds is definitely a win.

One question though: Won't we run into trouble with MIR inlining here as it potentially mixes in locals from other crates?

**Comment by eddyb** [comment]
> Won't we run into trouble with MIR inlining here as it potentially mixes in locals from other crates?

Where does MIR refer to variable bindings by some kind of ID? If it does that's a bug IMO.

**Comment by michaelwoerister** [comment]
r=me once rebased.

**Pull request merged**